src/App.tsx

```diff
@@ -1,24 +1,270 @@
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isTodoAdding, setIsTodoAdding] = useState(false);
+
+  const activeTodosCounter = todos.filter(todo => !todo.completed).length;
+  const completedTodosCounter = todos.filter(todo => todo.completed).length;
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const timerId = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerId.current) {
+      window.clearTimeout(timerId.current);
+    }
+
+    timerId.current = window.setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  }, [errorMessage]);
+
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filterField);
+  }, [todos, filterField]);
+
+  const handleDeleteTodo = useCallback((todoId: number) => {
+    setErrorMessage('');
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    return deleteTodo(todoId)
+      .then(() => {
+        setTodos((prevState) => (
+          prevState.filter(todo => todo.id !== todoId)
+        ));
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to delete a todo');
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  }, []);
+
+  const handleAddTodo = useCallback((newTodoTitle: string) => {
+    setErrorMessage('');
+    setTempTodo({
+      id: 0,
+      title: newTodoTitle,
+      userId: 0,
+      completed: false,
+    });
+
+    return addTodo(newTodoTitle.trim())
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to add a todo');
+        throw error;
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  }, []);
+
+  const handleUpdateTodo = (todo: Todo, newTodoTitle: string) => {
+    setErrorMessage('');
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    updateTodo({
+      id: todo.id,
+      title: newTodoTitle,
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch((error) => {
+        setErrorMessage('Unable to update a todo');
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleClearCompleted = () => {
+    const compeletedTodos = todos.filter(todo => todo.completed);
+
+    compeletedTodos.forEach(todo => {
+      handleDeleteTodo(todo.id);
+    });
```

Comment:
When you need to send a batch of requests it's better to create an array of requests and handle them with `Promise.all` or `Promise.allsettled`